sdk/webpubsub/azure-messaging-webpubsubclient/azure/messaging/webpubsubclient/aio/_client.py

```diff
@@ -0,0 +1,1091 @@
+        self._ack_map: AckMapAsync = AckMapAsync()
+        self._sequence_id = SequenceId()
+        self._state = WebPubSubClientState.STOPPED
+        self._ack_id = 0
+        self._url = None
+        self._ws: Optional[WebSocketAppAsync] = None
+        self._handler: Dict[str, List[Callable]] = {
+            CallbackType.CONNECTED: [],
+            CallbackType.DISCONNECTED: [],
+            CallbackType.REJOIN_GROUP_FAILED: [],
+            CallbackType.GROUP_MESSAGE: [],
+            CallbackType.SERVER_MESSAGE: [],
+            CallbackType.STOPPED: [],
+        }
+        self._last_disconnected_message: Optional[DisconnectedMessage] = None
+        self._connection_id: Optional[str] = None
+        self._is_initial_connected = False
+        self._is_stopping = False
+        self._last_close_event: Optional[CloseEvent] = None
+        self._reconnection_token: Optional[str] = None
+        self._task_seq_ack: Optional[asyncio.Task] = None
+        self._task_listen: Optional[asyncio.Task] = None
+        self._ack_timeout: float = ack_timeout
+        self._start_timeout: float = start_timeout
+        self._user_agent: Optional[str] = user_agent
+        self._logging_enable: bool = logging_enable
+
+    def _next_ack_id(self) -> int:
+        self._ack_id = self._ack_id + 1
+        return self._ack_id
+
+    async def _send_message(self, message: WebPubSubMessage, **kwargs: Any) -> None:
+        pay_load = self._protocol.write_message(message)
+        if self._ws.closed:  # type: ignore
+            raise SendMessageError("The websocket connection is not connected.")
+
+        await self._ws.send(pay_load)  # type: ignore
+        if kwargs.pop("logging_enable", False) or self._logging_enable:
+            _LOGGER.debug("\nconnection_id: %s\npay_load: %s", self._connection_id, pay_load)
+
+    async def _send_message_with_ack_id(
+        self,
+        message_provider: Callable[[int], WebPubSubMessage],
+        ack_id: Optional[int] = None,
+        **kwargs: Any,
+    ) -> None:
+        if ack_id is None:
+            ack_id = self._next_ack_id()
+
+        message = message_provider(ack_id)
+        # Unless receive ack message, we assume the message is not sent successfully.
+        if not self._ack_map.get(ack_id):
+            self._ack_map.add(
+                ack_id,
+                SendMessageErrorOptionsAsync(
+                    error_detail=AckMessageError(name="", message="Timeout while waiting for ack message.")
+                ),
+            )
+        try:
+            await self._send_message(message, **kwargs)
+        except Exception as e:
+            self._ack_map.pop(ack_id)
+            raise e
+
+        message_ack = self._ack_map.get(ack_id)
+        raise_for_empty_message_ack(message_ack, ack_id)
+
+        _LOGGER.debug("wait for ack message with ackId: %s", ack_id)
+        if message_ack:
+            await asyncio.wait_for(message_ack.event.wait(), timeout=self._ack_timeout)
```

Comment:
if event is not set within the timeout, `asyncio.wait_for` will raise an `asyncio.TimeoutError`.  Youll have to try/except for that here to handle that case.

Reply:
exception caused by`_send_message_with_ack_id` will be handled in `_retry` logic